AWS migration devnet and gemini

DaMandal0rian opened the pull request:
The PR migrates our devnet network and gemini infra from DO to AWS
- this is related to issue #122 
- start refactoring for #135

DaMandal0rian committed:
Devnet AWS

- migrate devnet from DO to AWS

DaMandal0rian committed:
Update images latest release

- use latest images for node keys

DaMandal0rian committed:
Add gemini-3d module

- add gemini-3d module

Minor changes:
- add ingress ports for firewall
- add DNS records output

DaMandal0rian committed:
Add EBS block support

- add ebs volumes and attachments to ec2 instances.
- Minor change OS user variables & ssh
- increase timeouts

DaMandal0rian committed:
Fixes to primitives

- fix dns label index
- fix apt package  prompts
- permissions and run in user-space
- remove node key generation local provisioner

vedhavyas reviewed (changes requested):
This is not the way to go. IIRC, we wanted to move away from current network primitives to something like ansible i believe. Maybe @nazar-pc could chime in here regarding that.

Having said that, if we still want to use network primitives there is no reason to copy and create a new set of same scripts. The network primitive's scripts and a good chunk of terraform scripts like provisioner are independent of the cloud. So there is no difference between Digital ocean or Heztner there once instances are created since we just need SSH access. If we want to integrate AWS and deprecate DO, then add the required variables exposed through the main module like DO for AWS.

DaMandal0rian committed:
Update .gitignore

DaMandal0rian committed:
Fix create compose scripts

- fix character escaping
- minor edits

DaMandal0rian committed:
Fixes primitives, scripts and variables

- add variables for volume type and size
- change directory permissions
- variable interpolation

DaMandal0rian committed:
Fixes to network and instances

- add copy for dsn_bootstrap_node keys
- add ingress port for farmer
- Fix ingress cidr for http/https
- add output for rpc
- increase ebs storage size on creation
- remove EBS volumes and attachments

nazar-pc commented on aws/devnet/main.tf:
```diff
@@ -0,0 +1,97 @@
+    node-dsn-port       = 30434
+    enable-domains      = true
+    domain-id           = var.domain_id
+    domain-labels       = var.domain_labels
```
Just out of curiosity, why are some things using `-` as separator and others `_`? Is this some kind of naming convention? Looks a bit inconsistent.

DaMandal0rian replied:
I use it to distinguish between config variables and the global variables. However, our older networks have it more inconsistent. i.e

```
bootstrap-node-config = {
    droplet_size        = var.droplet-size
    deployment-version  = 
    regions             = []
    nodes-per-region    = 0
}
```

nazar-pc commented on aws/devnet/variables.tf:
```diff
@@ -0,0 +1,122 @@
+  description = "Tag of the domain to run"
+  type        = list(string)
+  default     = ["evm", "payments"]
+}
```
A bit confusing that one has 2 elements and the other has 3

nazar-pc commented on aws/devnet/variables.tf:
```diff
@@ -0,0 +1,122 @@
+variable "bootstrap_node_ips" {
+  description = "IP of boostrap node"
+  type        = list(string)
+  default     = [""]
```
Do you need this default value? It makes no sense as an IP address, it'd better fail and request user to provide an IP instead of silently using invalid value.

DaMandal0rian replied:
These will be removed since we are strictly using AWS only, and have no need for these.

nazar-pc commented on aws/devnet/variables.tf:
```diff
@@ -0,0 +1,122 @@
+  default     = ["us-east-1a", "us-east-1b"]
+}
+
+variable "instance_count" {
+  type    = number
+  default = 1
+}
```
Is this per zone or global?

DaMandal0rian replied:
this is per zone.

nazar-pc commented on aws/devnet/variables.tf:
```diff
@@ -0,0 +1,122 @@
+variable "public_subnet_cidrs" {
+  type        = list(string)
+  description = "Public Subnet CIDR values"
+  default     = ["172.31.1.0/24"]
```
This isn't public though, what does this mean?

DaMandal0rian replied:
this is just the IP address range to assign the nic cards for machines in the subnet, public interfaces are exposed on another interface or through the internet gateway or elastic IPs.

nazar-pc replied:
Why is it called `public_subnet_cidrs` then if it is not public? I'm confused.

DaMandal0rian replied:
because that is the subnet that has public access to internet. The only thing that distinguishes pulic and private is that one has full ingress and egress and the private uses the public subnet nat gateway and internet gateway.  Please see https://docs.aws.amazon.com/vpc/latest/userguide/configure-subnets.html#subnet-basics

nazar-pc commented on aws/devnet/variables.tf:
```diff
@@ -0,0 +1,122 @@
+
+variable "disk_volume_size" {
+  type    = number
+  default = 50
```
I'd probably prefer to not have default value, but rather require this being specified for every node explicitly. It'll likely be different for different types of nodes anyway.

DaMandal0rian replied:
sure, will do that, I set these as a precaution of smaller values so mistakenly don't create large volumes and incur extra charges, the size of disks are set in the module configs and will overwrite this default.

nazar-pc replied:
If they are set explicitly anyway, why having a default?

DaMandal0rian replied:
I have removed the default. The variable will be set in tfvars file to desired amount needed.

DaMandal0rian replied:
When I remove it, we get the following:

│ Error: Missing required argument
│ 
│   on main.tf line 1, in module "devnet":
│    1: module "devnet" {
│ 
│ The argument "disk_volume_size" is required, but no definition was found.

Again, if the user needs to adjust this they can set the desired amount in the module itself if they need something higher than the default, or I'm happy setting a higher default.

nazar-pc replied:
Well, this is exactly the problem I am trying to explain to you. Since this is a default value, it should not be required. If something else doesn't work after you remove default value, it seems to indicate you were in fact using default value, which as we already discussed is way too low and should be multiple times higher and specified explicitly.

nazar-pc commented on aws/network-primitives/bootstrap_node_provisioner.tf:
```diff
@@ -0,0 +1,153 @@
+locals {
+  bootstrap_nodes_ip_v4 = flatten([
+    #   [var.bootstrap-node-config.additional-node-ips],
+    [aws_instance.bootstrap_node.*.public_ip]
```
Something left from older attempts

DaMandal0rian replied:
No, i just temporarily commented it, but it will be removed since we are not using another provider with this and using this strictly for AWS.

nazar-pc commented on aws/network-primitives/bootstrap_node_provisioner.tf:
```diff
@@ -0,0 +1,153 @@
+  provisioner "file" {
+    source      = "./dsn_bootstrap_node_keys.txt"
+    destination = "/subspace/dsn_bootstrap_node_keys.txt"
+  }
```
I find it strange to copy all the keys to all nodes, which are not even removed afterwards. Even when node is pruned, the keys are still present and might get mixed with something later.

DaMandal0rian replied:
This is remnants from the older version, I will modify the prune script to remove keys, but currently it would just copy over the existing ones.

nazar-pc replied:
Can we add a TODO then? It is easy to forget and others don't have that context in the first place.

DaMandal0rian replied:
Already have, and have added to the prune script already to remove key files. This was done in the prune_docker_system.sh script file.

nazar-pc commented on aws/network-primitives/bootstrap_node_provisioner.tf:
```diff
@@ -0,0 +1,153 @@
+      "sudo /subspace/create_compose_file.sh ${var.bootstrap-node-config.reserved-only} ${length(local.bootstrap_nodes_ip_v4)} ${count.index}",
+
+      # start subspace node
+      "sudo chown ubuntu:ubuntu /subspace/docker-compose.yml",
```
Isn't it already owned by the user that runs it?

DaMandal0rian replied:
It's just a precaution to ensure the ownership is set.

nazar-pc replied:
Hm... it makes little sense to me still. You've created it under current user, then using `sudo` to set ownership to current user :man_shrugging: 

It is either necessary or not, I don't like arbitrary "just in case" code that is not actually required.

DaMandal0rian replied:
It is 100% necessary. Otherwise it's owned by root user.

`-rw-r--r--  1 root   root   2050 Jun  9 07:10 docker-compose.yml`

nazar-pc replied:
Right, maybe remove `sudo` from above two commands instead? It seems to be happening backwards because of that.

DaMandal0rian replied:
It fails in that case

nazar-pc replied:
The natural response to that would be how does it fail exactly?

What I see is that it reads some environment variables and writes text to a file. None of those things inherently require root access. I guess the issue might be if `/subspace` is created and owned by root user, but then why would you want the file in it to be owned by `ubuntu` user I do not understand.

DaMandal0rian replied:
`provisioner "file"` uses scp to copy the files. It fails with permission denied for the user `ubuntu`  since the path  `/subspace` is directly under the `/` path.

https://github.com/hashicorp/terraform/issues/8238 

Alternatively, I can copy to a /home/ubuntu/subspace because the user `ubuntu` will have the right uid/gid set.

nazar-pc commented on aws/network-primitives/bootstrap_node_provisioner.tf:
```diff
@@ -0,0 +1,153 @@
+
+      # start subspace node
+      "sudo chown ubuntu:ubuntu /subspace/docker-compose.yml",
+      "sudo docker compose -f /subspace/docker-compose.yml up -d --build ",
```
What are we building here? Generated `docker-compose.yml` doesn't seem to contain anything that needs to be built.

This seems to be the same kind of thing as installing various unused packages. Your code is also your documentation. When I see `--build`, I immediately start wondering why do we build anything just to discover that we in fact do not build anything.

nazar-pc replied:
Also why do we start the container with `docker-compose` if that should be done by systemd instead?

DaMandal0rian replied:
I'd prefer to not use systemd with terraform since it can break the run and would rather use systemd with configuration management. The --build flag just forces a rebuild of the images, not really applicable here since we don't use the `latest` tag, but it just forces a rebuild of the images before running them.

nazar-pc replied:
If there is nothing to rebuild, it should not specify `--build`.

As to using systemd or not, since we use systemd right now, this line is incorrect since it will start container without systemd and will not apply iptables rules as the result.

DaMandal0rian replied:
The Iptables and systemd will not be configured in terraform, this is not the way to go. Also, will those Iptable rules have effect on the aws instances?

`docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 172.17.255.255

ens5: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9001
        inet 172.31.1.139  netmask 255.255.255.0  broadcast 172.31.1.255`
        
   Secondly, those rules might be for older docker versions, but as of lately they recommend not to manually manipulate the rules.
   https://docs.docker.com/network/packet-filtering-firewalls/#add-iptables-policies-before-dockers-rules

Thirdly, the current rules do nothing since they are targeting the wrong chain and should be targeting `DOCKER-USER`

nazar-pc replied:
I thought we are deploying the same thing we had with minimal changes to AWS. What you're describing is quite far from it. Moreover, you're saying it is not the way to go, yet this PR still deploys the same systemd service and iptables rules, but in contrast to deployment to Hetzner where everything was done properly and in correct order, the way you wrote it is not equivalent, hence this comment and request to correct it to the way it was before.

The rules will certainly have an effect. They are probably redundant, but I have not visibility into what you have actually changed here comparing to what we had before unless I check out code locally and do file comparison manually.

> Thirdly, the current rules do nothing since they are targeting the wrong chain and should be targeting DOCKER-USER

They were certainly working properly and were not related to Docker as such at all. Our servers were blocked on Hetzner because we were making requests to private IP ranges via public interfaces, that is what those rules were blocking and they certainly did the job.

nazar-pc commented on aws/network-primitives/instances.tf:
```diff
@@ -0,0 +1,297 @@
+
+  provisioner "remote-exec" {
+    inline = [
+      "sleep 180",
```
:thinking: why? This is the kind of thing that typically randomly fails in practice.

DaMandal0rian replied:
It gives the cloud-init scripts time to finish before trying to run the remote provisioners else it will timeout, the aws launches go though an initialization stage before the health checks are done and they system state is "ready/running".

nazar-pc replied:
Is there a way to detect when cloud-init finishes rather than having arbitrary timeouts?

180 seconds is A LOT, would be much better to proactively detect this rather than waiting for whole 3 minutes unnecessarily.

DaMandal0rian replied:
There is no way for cloud-init to tell terraform it's finished. However, I have removed this and just increased the timeout for ssh connections.

nazar-pc replied:
> There is no way for cloud-init to tell terraform it's finished. However, I have removed this and just increased the timeout for ssh connections.

It doesn't need to. `cloud-init` is doing something, right?  So you can have a loop here that will check for whatever you expect cloud-init to finish doing every few seconds and move forward only when it is done.

Removing delay and increasing timeout makes even less sense now. Will it automatically retry or something?

nazar-pc commented on aws/network-primitives/instances.tf:
```diff
@@ -0,0 +1,297 @@
+  ebs_optimized = true
+  ebs_block_device {
+    device_name = "/dev/sda1"
+    volume_size = 100
```
These disk sizes look severely underestimated. Are you sure they are sufficient?

DaMandal0rian replied:
actually, I will set these in configs.

nazar-pc replied:
Well, you did, but as far as I see you have used those 50G sized that I complained about in the other comment. So we have kind of made a full circle now. Or am I missing something and those values come from somewhere else?

DaMandal0rian replied:
Actually,  the comment is outdated, not sure why you're still seeing that.

nazar-pc replied:
This comment is outdated, but the issue is not, which is why I'm continuing this conversation. I see several places and none of them have the correct disk size. So to the best of my understanding we still have an issue even though code has changed since I wrote the first comment here.

nazar-pc commented on aws/network-primitives/network.tf:
```diff
@@ -0,0 +1,193 @@
+resource "aws_vpc" "gemini_vpc" {
```
Should it be called `gemini_*` if it is in primitives (here and below)?

DaMandal0rian replied:
That is the name of the resource, the naming is correct and can't be set dynamically afaik.

nazar-pc replied:
I'd argue the name is not correct. This file is `aws/network-primitives/network.tf`, meaning it is used by all networks as a primitive, not just by Gemini networks, so should be named accordingly, otherwise looks misleading to me.

DaMandal0rian replied:
I just used "network" instead, and the "name" tag will set the actual network name to identify it.

nazar-pc commented on aws/network-primitives/provider.tf:
```diff
@@ -0,0 +1,32 @@
+
+    cloudflare = {
+      source  = "cloudflare/cloudflare"
+      version = "~> 3.18.0"
```
What is the reason to not upgrade minor version of CloudFlare dependency?

DaMandal0rian replied:
the `~>` will pull in the latest versions greater than 3.18.0. I have upgraded the provider version.

nazar-pc commented on aws/network-primitives/network.tf:
```diff
@@ -0,0 +1,193 @@
+  route_table_id = element(aws_route_table.private_route_table.*.id, count.index)
+}
+
+resource "aws_security_group" "gemini_sg" {
```
Looks like port 50000 is missing in this list despite used in DSN bootstrap node container

DaMandal0rian replied:
Will add this, actually now i think of it that port was never set in firewall from our DO declarations.

nazar-pc commented on aws/network-primitives/scripts/install_docker.sh:
```diff
@@ -0,0 +1,25 @@
+# updates
+export DEBIAN_FRONTEND=noninteractive
+sudo apt update -y
+sudo apt install -y curl jq
```
`jq` doesn't seem to be used anywhere

nazar-pc reviewed (commented):
Commit structure is a bit of a mess, I'd like to request you to re-read https://www.notion.so/subspacelabs/Pull-Request-Guide-cd8d0f0f5fe643d5bfde6145450d5ff6 (I have recently updated it, you likely didn't see the latest revision). It gives numerous suggestions how to make reviewers life easier that could be applied to this PR.

I left mostly questions, there was a lot of code duplication, so I might have missed something, but overall makes sense to me.

As to discussion about provisioning method and scripts.

I do agree that there was no need to copy-paste scripts. You can add `sudo` and it will continue working the same way on DO and Hetzner just fine. As to the user it also doesn't make a meaningful difference and container can be stared under `root`, permissions are lower in the container anyway.

So you count have unified them and that is possible to do in a separate PR such that diff here is smaller. I don't see a big reason to change them in a significant way even though we could simplify things for AWS/DO, systemd is only needed for Hetzner. The deployment method will change in the future, but there is no need to do all changes at once.

That said, please don't change to shared scripts in this PR, let's leave it as is, it'll require even more time to review thousands of lines of code. Just take this as a feedback for future changes.

DaMandal0rian committed:
change images and backend

DaMandal0rian committed:
Fixes to networks + latest images

nazar-pc reviewed (commented):
> I understand, I put the scripts into the new network primitives folder under the AWS folder, most of them copied over, and only 2 were actually modified. I didn't modify the others, the diff is just them being added to the path.
>
> Additionally, i didn't want to spend too much time on these scripts since they will be removed in the future completely.

Understood, in this case a good practice is to copy them with 0 modifications in a separate commit (such that it is clear to reviewer that it doesn't need to be reviewed) and then doing modifications in a distinct commit, such that reviewer will spend little time if they are already familiar with the previous version. Copying and modifying at the same time means the whole file needs to be re-reviewed from scratch.

> These were just migrated over for now so we can launch faster, and will be removed completely in the future. If we are trying to seek perfection in every detail beforehand, rather than a work-in-progress with iterations, then it's going to just take longer to deliver things imo. Infra always changes and has improvements along the way.

Completely agree with you, this is why we need a separate commit that literally says something like "Files are copied from X with zero modifications, will be changes in the next commit" and maybe even add an explanation why did you decide to copy the files rather than reusing originals. What you explained makes sense to me, but it was not clear from the original PR, commits and description provided.

nazar-pc reviewed (commented):
Please don't resolve every single comment just because you have replied to it. I have to expand every single one, reply to it and unresolve it manually after that. Only resolve those you're 100% were resolved and no questions there will remain. I will resolve those that I have no question about anymore myself.

DaMandal0rian committed:
set devnet genesis hash and set latest snapshot

DaMandal0rian committed:
EBS volume performance

- use gp3 disk types
- increase iops
- increase throughput

DaMandal0rian committed:
Check cloud init finished before running scripts

DaMandal0rian commented on aws/network-primitives/domain_node_provisioner.tf:
```diff
@@ -0,0 +1,197 @@
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mkdir -p /subspace",
+      "sudo mkdir -p /home/ubuntu/bin/",
```
remove this

DaMandal0rian commented on aws/network-primitives/farmer_node_provisioner.tf:
```diff
@@ -0,0 +1,157 @@
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mkdir -p /subspace",
+      "sudo mkdir -p /home/ubuntu/bin/",
```
remove

DaMandal0rian commented on aws/network-primitives/rpc_node_provisioner.tf:
```diff
@@ -0,0 +1,193 @@
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mkdir -p /subspace",
+      "sudo mkdir -p /home/ubuntu/bin",
```
remove

DaMandal0rian commented on aws/network-primitives/full_node_provisioner.tf:
```diff
@@ -0,0 +1,154 @@
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mkdir -p /subspace",
+      "sudo mkdir -p /home/ubuntu/bin/",
```
remove

DaMandal0rian commented on aws/network-primitives/bootstrap_node_provisioner.tf:
```diff
@@ -0,0 +1,152 @@
+  provisioner "remote-exec" {
+    inline = [
+      "sudo mkdir -p /subspace",
+      "sudo mkdir -p /home/ubuntu/bin/",
```
remove

DaMandal0rian committed:
Improvements to provisioners

- run scripts and copy files in user-space
- use cloud-init wait for instances to wait for cloud-init to complete
- fix permissions and ownership to user

DaMandal0rian committed:
remove unused scripts

DaMandal0rian committed:
Remove default volume size + add examples file

DaMandal0rian committed:
Fix paths in scripts

- remove systemd in scripts

nazar-pc reviewed (approved):
Latest version makes much more sense, thanks!
